include/exadg/vector_tools/arbitrary_solution_transfer.h

```diff
@@ -0,0 +1,164 @@
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_VECTOR_TOOLS_ARBITRARY_SOLUTION_TRANSFER_H
+#define INCLUDE_VECTOR_TOOLS_ARBITRARY_SOLUTION_TRANSFER_H
+
+#include <deal.II/base/mpi_remote_point_evaluation.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/fe/mapping.h>
+#include <deal.II/numerics/vector_tools.h>
+
+namespace ExaDG
+{
+template<int dim>
+class ArbitrarySolutionTransfer
+{
+public:
+  /**
+   * Set source and target triangulations on which solution transfer is performed.
+   *
+   * @param[in] dof_handler_dst_in Target DofHandler.
+   * @param[in] mapping_dst_in Target Mapping.
+   * @param[in] dof_handler_src_in Source DofHandler.
+   * @param[in] mapping_src_in Source Mapping.
+   */
+  void
+  reinit(dealii::DoFHandler<dim> const & dof_handler_dst_in,
+         dealii::Mapping<dim> const &    mapping_dst_in,
+         dealii::DoFHandler<dim> const & dof_handler_src_in,
+         dealii::Mapping<dim> const &    mapping_src_in)
+  {
+    dof_handler_dst = &dof_handler_dst_in;
+    dof_handler_src = &dof_handler_src_in;
+
+    rpe.reinit(collect_mapped_support_points(dof_handler_dst_in, mapping_dst_in),
+               dof_handler_src_in.get_triangulation(),
+               mapping_src_in);
+  }
+
+  /**
+   * Transfer the solution between source and target triangulations, that were set with reinit.
+   * At support points which are not overlapping the source triangulation, corresponding dof entries
+   * are set to 0.
+   *
+   * @param[in] dst Target Dof Vector.
+   * @param[in] src Source Dof Vector.
+   */
+  template<int n_components, typename VectorType1, typename VectorType2>
+  void
+  transfer_solution(VectorType1 &                                               dst,
+                    VectorType2 const &                                         src,
+                    dealii::VectorTools::EvaluationFlags::EvaluationFlags const flags =
+                      dealii::VectorTools::EvaluationFlags::avg) const
+  {
+    src.update_ghost_values();
+    auto const values =
+      dealii::VectorTools::point_values<n_components>(rpe, *dof_handler_src, src, flags);
+    src.zero_out_ghost_values();
+
+    fill_dof_vector_with_values<n_components>(dst, *dof_handler_dst, values);
```

Comment:
I thought I have written another comment, but I can not find it anymore ... so I write it again: For me, it looks like the two functions `collect_mapped_support_points()` and `fill_dof_vector_with_values()` assume that we go through the mesh/cells in the same order. Is this implementation safe enough? Put differently: Who takes care of the correct indexing so that it is guaranteed that the `values` returned by `VectorTools::point_values()` are written correctly to the dof-vector? It looks like we are just looping through the vector from begin to end, so there must be some implicit convention behind, right? Shouldn't we explicitly use maps with cell_ids/dof_indices as we use in `InterfaceCoupling`?

Reply:
The convention of `RemotePointEvaluation` and `VectorTools::point_values` is that the order of the returned values be the same as the one in points passed in. I did look at the code and it seems that this is fulfilled in terms of the order of loops. I think the usage should be fine. Clearly, there is still some small ambiguity in terms of the loop done in the two functions `collect_mapped_support_points` and `fill_dof_vector_with_values`, but I do not see an immediate benefit of the functions we have in `InterfaceCoupling`. The argument is that lookup via `std::map` is (probably) rather slow for a volume quantity, and we still rely on the convention that we look up points previously inserted, so I would primarily see this as beneficial during code development, not for catching run-time errors.

Reply:
ok, so let's use the current approach. I have one short question, though: There is probably no generic deal.II functionality to loop only over the locally owned cells so that we do not have to write the if-statement with `continue` in the code here? I wonder how often these lines of code are written here and there.

Reply:
@nfehn You can use filter, like here: https://github.com/dealii/dealii/blob/dd99ecfe16df9287195726e4444bb448a5b78fe5/examples/step-85/step-85.cc#L363-L365 (personally, I don't like).

Reply:
Indeed, I also personally prefer writing `if` statements over these iterator filters, but I am not sure if it more a habit or real preference. A more direct use (from `mg_global_coarsening.templates.h`) is:
```
      for (const auto &cell : dof_handler_fine.active_cell_iterators() |
                                IteratorFilters::LocallyOwnedCell())
        is_locally_owned_fine.add_index(cell_id_translator.translate(cell));
```